sklearn/utils/validation.py

```diff
@@ -445,6 +445,12 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
         # not a data type (e.g. a column named dtype in a pandas DataFrame)
         dtype_orig = None
 
+    # check if the object contains several dtypes (typically a pandas
+    # DataFrame), and store them. If not, store None.
+    dtypes_orig = None
+    if hasattr(array, "dtypes"):
```

Comment:
Maybe we should be a bit more precise on the duck typing, to avoid to capture other objects that happen by chance to have a dtype attribute. For instance, we could test for "__array__" which, I believe is the method used by numpy to convert to a numpy array.

Reply:
Note that the attribute here is dtypes, not dtype.

Reply:
Thanks, will do